File: chapt2/demo_cpp_pkg/src/learn_thread.cpp
#include <iostream>
#include <chrono>
#include <thread>
#include <functional>
#include <cpp-httplib/httplib.h>

class DownLoad
{
public:
    void download(const std::string &host, const std::string &path,
                  const std::function<void(const std::string &, const std::string &)> &callback)
    {
        std::cout << "线程ID:" << std::this_thread::get_id() << std::endl;
        httplib::Client client(host);
        auto response = client.Get(path);
        if (response && response->status == 200)
        {
            callback(path, response->body);
        }
        else
        {
            std::cout << "下载错误" << std::endl;
            std::cout << "status:" << response->status << std::endl;
        }
    }

    void start_download(const std::string &host, const std::string &path,
                        const std::function<void(const std::string &, const std::string &)> &callback)
    {
        using namespace std::placeholders;
        auto download_func = std::bind(&DownLoad::download, this, _1, _2, _3);
        std::thread download_thread(download_func, host, path, callback);
        download_thread.detach();
    }
};

int main(int argc, char **agrv)
{
    DownLoad d;
    auto callback_func = [](const std::string &path, const std::string &res) -> void
    { std::cout << "下载完成：" << path << " 共 " << res.length() << "字 , 内容为:" << res.substr(0, 5) << std::endl; };

    d.start_download("http://0.0.0.0:8000", "/nove11.txt", callback_func);
    d.start_download("http://0.0.0.0:8000", "/nove12.txt", callback_func);
    d.start_download("http://0.0.0.0:8000", "/nove13.txt", callback_func);

    std::this_thread::sleep_for(std::chrono::microseconds(1000 * 10));

    return 0;
}

File: chapt2/demo_python_pkg/demo_python_pkg/learn_thread.py
import threading
import requests

class DownLoad:
    def download(self,url,callback):
        print(f'线程：{threading.get_ident()},开始下载:{url}')
        response = requests.get(url)
        response.encoding = 'UTF-8'
        print(f'内容：{response.text}')
        callback(url,response.text)
        
    def start_download(self,url,callback):
        thread = threading.Thread(target=self.download,args=(url,callback))
        thread.start()
        
def download_finish_callback(url,result):
        print(f'{url}:下载完成，共{len(result)}字，内容为：{result[:5]}…')
        
def main():
    d = DownLoad()
    d.start_download('http://0.0.0.0:8000/nove11.txt',download_finish_callback)
    d.start_download('http://0.0.0.0:8000/nove12.txt',download_finish_callback)
    d.start_download('http://0.0.0.0:8000/nove13.txt',download_finish_callback)